### PR TITLE
Add link_read and link_write to MachineController

### PR DESCRIPTION
**Pull request by mossblaser**
These methods allow users to send PEEK/POKE requests to routers of neighbouring
chips. The implementation is not performance-optimised since this feature
expected to be used in a debug setting.

### CODE REVIEW

File: rig/machine_control/machine_controller.py [hunk 1]
```diff
@@ -396,6 +396,127 @@ def read(self, address, length_bytes, x, y, p=0):
...
+        This method allows you to instruct a monitor processor to send 'POKE'
+        nearest-neighbour packets to a neighbouring chip. These packets are
+        handled directly by the SpiNNaker router in the neighbouring chip,
+        potentially allowing advanced debug or recovery of a chip rendered
+        otherwise unreachable.
+
+        Parameters
+        ----------
+        address : int
+            The address at which to start reading the data. Only addresses in
```
Comment (mundya) [inline rig/machine_control/machine_controller.py]: Write, surely?

File: rig/machine_control/machine_controller.py [hunk 2]
```diff
@@ -396,6 +396,127 @@ def read(self, address, length_bytes, x, y, p=0):
         return connection.read(self.scp_data_length, self.scp_window_size,
                                x, y, p, address, length_bytes)
 
+    @ContextMixin.use_contextual_arguments()
+    def link_write(self, address, data, x, y, link):
```
Comment (mundya) [inline rig/machine_control/machine_controller.py]: How about `write_across_link`?

File: rig/machine_control/machine_controller.py [hunk 3]
```diff
@@ -396,6 +396,127 @@ def read(self, address, length_bytes, x, y, p=0):
...
+            SCP link-write commands.
+        x : int
+        y : int
+            The coordinates of the chip from which the command will be sent,
+            *not* the coordinates of the chip on which the write will be
+            performed.
+        link : :py:class:`rig.machine.Links`
+            The link down which the write should be sent.
+        """
+        if address & 0b11:
```
Comment (mundya) [inline rig/machine_control/machine_controller.py]: I think using % 4 would be clearer.

File: rig/machine_control/machine_controller.py (hunk 1 above)
Comment (mossblaser) [inline rig/machine_control/machine_controller.py]: Of course ;) For some reason GitHub doesn't want to hide this comment... Should be fixed now!

**Comment by mossblaser** [comment]
Thanks for the review, @mundya! Everything should now be up-to-scratch!

File: rig/machine_control/machine_controller.py [hunk 4]
```diff
@@ -396,6 +396,127 @@ def read(self, address, length_bytes, x, y, p=0):
...
+    @ContextMixin.use_contextual_arguments()
+    def write_across_link(self, address, data, x, y, link):
+        """Write a bytestring to an address in memory on a neigbouring chip.
+
+        .. warning::
+
+            This function is intended for low-level debug use only and is not
+            optimised for performance nor intended for more general use.
+
+        This method allows you to instruct a monitor processor to send 'POKE'
```
Comment (mundya) [inline rig/machine_control/machine_controller.py]: "This method instructs"? - Being very picky...

File: rig/machine_control/machine_controller.py [hunk 5]
```diff
@@ -396,6 +396,127 @@ def read(self, address, length_bytes, x, y, p=0):
...
+            The coordinates of the chip from which the command will be sent,
+            *not* the coordinates of the chip on which the write will be
+            performed.
+        link : :py:class:`rig.machine.Links`
+            The link down which the write should be sent.
+        """
+        if address % 4:
+            raise ValueError("Addresses must be word-aligned.")
+        if len(data) % 4:
+            raise ValueError("Data must be a multiple of words in length.")
```
Comment (mundya) [inline rig/machine_control/machine_controller.py]: "Data must be a whole number of words."?

**Comment by mundya** [comment]
Aside from being increasingly curmudgeonly this LGTM

**Comment by mundya** [comment]
LGTM, merge when ready :)

**Pull request merged**